File: src/svg_to_sfsymbol/phase5.py
"""
Phase 5: merge generated weight×size SVGs into Apple's SF Symbol square template.

For each ``<g id="Black-L">`` … under ``<g id="Symbols">``, replace the
``SFSymbolsPreviewWireframe`` path with the corresponding ``{id}.svg`` content,
uniformly scaled and centered to the wireframe path's bounding box in local
coordinates, then remove the wireframe.
"""

from __future__ import annotations

import copy
import warnings
from pathlib import Path
from typing import Literal, Optional, Tuple
from xml.etree import ElementTree as ET

from svg.path import parse_path

from svg_to_sfsymbol.phase1 import element_to_bytes, parse_svg_file, read_view_box
from svg_to_sfsymbol.phase2 import PHASE2_SIZE_SUFFIX, PHASE2_WEIGHTS
from svg_to_sfsymbol.phase3 import _local_tag, parse_weight_stem

SVG_NS = "http://www.w3.org/2000/svg"
_G_TAG = f"{{{SVG_NS}}}g"
_PATH_TAG = f"{{{SVG_NS}}}path"
_DEFS_TAG = f"{{{SVG_NS}}}defs"
WIREFRAME_CLASS = "SFSymbolsPreviewWireframe"

MissingPolicy = Literal["skip", "warn", "fail"]


def default_square_template_path() -> Optional[Path]:
    """
    Return ``resources/square_template.svg`` next to the project root that contains
    ``src/svg_to_sfsymbol``, if that file exists.

    Used by ``convert`` to run phase 5 without extra flags. Returns ``None`` when
    the file is absent (e.g. a minimal install without bundled resources).
    """
    here = Path(__file__).resolve().parent
    candidate = here.parent.parent / "resources" / "square_template.svg"
    return candidate if candidate.is_file() else None


def _expected_slot_ids() -> frozenset[str]:
    return frozenset(
        f"{w}-{suf}"
        for w in PHASE2_WEIGHTS
        for _name, suf in PHASE2_SIZE_SUFFIX
    )


def _find_symbols_group(root: ET.Element) -> ET.Element:
    for el in root.iter():
        if _local_tag(el.tag) == "g" and el.get("id") == "Symbols":
            return el
    raise ValueError('No <g id="Symbols"> found in template.')


def _ensure_template_defs(svg_root: ET.Element) -> ET.Element:
    for child in svg_root:
        if _local_tag(child.tag) == "defs":
            return child
    defs = ET.Element(_DEFS_TAG)
    svg_root.insert(0, defs)
    return defs


def _wireframe_bbox(d: str) -> Tuple[float, float, float, float]:
    """Axis-aligned bbox (minx, miny, width, height) of path ``d``."""
    path = parse_path(d)
    x_min, y_min, x_max, y_max = path.boundingbox()
    bw = x_max - x_min
    bh = y_max - y_min
    if bw <= 0 or bh <= 0:
        raise ValueError(f"Degenerate wireframe bbox from path d={d[:80]!r}…")
    return (x_min, y_min, bw, bh)


def _partition_icon_children(
    icon_root: ET.Element,
) -> tuple[list[ET.Element], list[ET.Element]]:
    defs_nodes: list[ET.Element] = []
    other: list[ET.Element] = []
    for child in list(icon_root):
        if _local_tag(child.tag) == "defs":
            defs_nodes.append(child)
        else:
            other.append(child)
    return defs_nodes, other


def _class_has_wireframe(class_attr: Optional[str]) -> bool:
    if not class_attr:
        return False
    return WIREFRAME_CLASS in class_attr.split()


def _merge_icon_defs_into(
    template_defs: ET.Element, defs_elements: list[ET.Element]
) -> None:
    existing_ids = {
        el.get("id")
        for el in template_defs.iter()
        if el.get("id") is not None
    }
    for defs_el in defs_elements:
        for child in list(defs_el):
            cid = child.get("id")
            if cid is not None and cid in existing_ids:
                continue
            template_defs.append(copy.deepcopy(child))
            if cid is not None:
                existing_ids.add(cid)


def _fill_slot(
    slot: ET.Element,
    icon_path: Path,
    template_defs: ET.Element,
) -> None:
    wire: Optional[ET.Element] = None
    for child in list(slot):
        if _local_tag(child.tag) != "path":
            continue
        if _class_has_wireframe(child.get("class")):
            wire = child
            break
    if wire is None:
        raise ValueError(
            f'Slot <g id="{slot.get("id")}"> has no {WIREFRAME_CLASS!r} path.'
        )

    d = wire.get("d")
    if not d:
        raise ValueError(f'Wireframe path in slot {slot.get("id")!r} has no d.')

    bx, by, bw, bh = _wireframe_bbox(d)

    icon_root = parse_svg_file(icon_path)
    _ox, _oy, sw, sh = read_view_box(icon_root)
    if sw <= 0 or sh <= 0:
        raise ValueError(f"Invalid icon viewBox dimensions in {icon_path}")

    scale = min(bw / sw, bh / sh)
    tx = bx + (bw - scale * sw) / 2.0
    ty = by + (bh - scale * sh) / 2.0

    defs_nodes, visual = _partition_icon_children(icon_root)
    if defs_nodes:
        _merge_icon_defs_into(template_defs, defs_nodes)

    slot.remove(wire)

    wrap = ET.Element(_G_TAG)
    wrap.set("transform", f"translate({tx} {ty}) scale({scale})")
    for node in visual:
        wrap.append(copy.deepcopy(node))
    slot.append(wrap)


def run_phase5(
    template_path: Path,
    icons_dir: Path,
    output_path: Path,
    *,
    missing: MissingPolicy = "skip",
) -> dict[str, int | list[str]]:
    """
    Merge icons from ``icons_dir`` (files ``Black-L.svg``, …) into ``template_path``.

    ``missing``: ``skip`` leaves the wireframe; ``warn`` emits ``warnings.warn``;
    ``fail`` raises ``FileNotFoundError`` on the first missing icon for a template slot.

    Returns ``{"filled": n, "skipped": n, "missing_ids": [...]}``.
    """
    template_path = template_path.expanduser().resolve()
    icons_dir = icons_dir.expanduser().resolve()
    output_path = output_path.expanduser().resolve()

    if not template_path.is_file():
        raise FileNotFoundError(f"Template not found: {template_path}")
    if not icons_dir.is_dir():
        raise NotADirectoryError(f"Icons directory not found: {icons_dir}")

    root = parse_svg_file(template_path)
    symbols = _find_symbols_group(root)
    template_defs = _ensure_template_defs(root)
    expected = _expected_slot_ids()

    filled = 0
    skipped = 0
    missing_ids: list[str] = []

    for slot in list(symbols):
        if _local_tag(slot.tag) != "g":
            continue
        sid = slot.get("id")
        if not sid or sid not in expected:
            continue
        if parse_weight_stem(sid) is None:
            continue

        icon_file = icons_dir / f"{sid}.svg"
        if not icon_file.is_file():
            missing_ids.append(sid)
            if missing == "fail":
                raise FileNotFoundError(
                    f"Missing icon for template slot {sid!r}: {icon_file}"
                )
            if missing == "warn":
                warnings.warn(
                    f"Phase 5: missing icon for slot {sid!r} ({icon_file}), "
                    "leaving wireframe.",
                    stacklevel=2,
                )
            skipped += 1
            continue

        _fill_slot(slot, icon_file, template_defs)
        filled += 1

    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_bytes(element_to_bytes(root))

    return {"filled": filled, "skipped": skipped, "missing_ids": missing_ids}

File: server/src/phase5.ts
import { existsSync, writeFileSync, mkdirSync } from "node:fs";
import type { PathLike } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import SvgPath from "svgpath";
import { readViewBox } from "./phase1.js";
import { PHASE2_SIZE_SUFFIX, PHASE2_WEIGHTS } from "./phase2.js";
import { parseWeightStem } from "./phase3.js";
import {
  elementChildren,
  elementToBytes,
  localTag,
  parseSvgFile,
  SVG_NS,
  svgDocumentElement,
  type SvgDocument,
  type SvgElement,
} from "./xml.js";

const WIREFRAME_CLASS = "SFSymbolsPreviewWireframe";

export type MissingPolicy = "skip" | "warn" | "fail";

function repoRootFromModule(): string {
  const here = dirname(fileURLToPath(import.meta.url));
  return join(here, "..", "..");
}

export function defaultSquareTemplatePath(): string | null {
  const candidate = join(repoRootFromModule(), "resources", "square_template.svg");
  return existsSync(candidate) ? candidate : null;
}

function expectedSlotIds(): Set<string> {
  const s = new Set<string>();
  for (const w of PHASE2_WEIGHTS) {
    for (const [, suf] of PHASE2_SIZE_SUFFIX) {
      s.add(`${w}-${suf}`);
    }
  }
  return s;
}

function findSymbolsGroup(root: SvgElement): SvgElement {
  const walk = (el: SvgElement): SvgElement | null => {
    if (localTag(el) === "g" && el.getAttribute("id") === "Symbols") return el;
    for (const c of elementChildren(el)) {
      const f = walk(c);
      if (f) return f;
    }
    return null;
  };
  const g = walk(root);
  if (!g) throw new Error('No <g id="Symbols"> found in template.');
  return g;
}

function ensureTemplateDefs(svgRoot: SvgElement): SvgElement {
  for (const ch of elementChildren(svgRoot)) {
    if (localTag(ch) === "defs") return ch;
  }
  const doc = svgRoot.ownerDocument;
  if (!doc) throw new Error("template svg has no owner document");
  const defs = doc.createElementNS(SVG_NS, "defs");
  svgRoot.insertBefore(defs, svgRoot.firstChild);
  return defs;
}

function pathBBoxFromD(d: string): [number, number, number, number] {
  let minX = Infinity;
  let minY = Infinity;
  let maxX = -Infinity;
  let maxY = -Infinity;
  const add = (x: number, y: number) => {
    minX = Math.min(minX, x);
    minY = Math.min(minY, y);
    maxX = Math.max(maxX, x);
    maxY = Math.max(maxY, y);
  };
  let p: InstanceType<typeof SvgPath>;
  try {
    p = new SvgPath(d).abs().unshort().unarc();
  } catch {
    throw new Error(`Degenerate wireframe bbox: unparseable d`);
  }
  let cx = 0;
  let cy = 0;
  for (const s of (p as unknown as { segments: (string | number)[][] }).segments) {
    const cmd = String(s[0]);
    if (cmd === "M" || cmd === "m") {
      cx = Number(s[1]);
      cy = Number(s[2]);
      add(cx, cy);
    } else if (cmd === "L" || cmd === "l" || cmd === "T" || cmd === "t") {
      cx = Number(s[s.length - 2]);
      cy = Number(s[s.length - 1]);
      add(cx, cy);
    } else if (cmd === "H" || cmd === "h") {
      cx = Number(s[1]);
      add(cx, cy);
    } else if (cmd === "V" || cmd === "v") {
      cy = Number(s[1]);
      add(cx, cy);
    } else if (cmd === "C" || cmd === "c") {
      add(Number(s[1]), Number(s[2]));
      add(Number(s[3]), Number(s[4]));
      cx = Number(s[5]);
      cy = Number(s[6]);
      add(cx, cy);
    } else if (cmd === "Q" || cmd === "q") {
      add(Number(s[1]), Number(s[2]));
      cx = Number(s[3]);
      cy = Number(s[4]);
      add(cx, cy);
    } else if (cmd === "Z" || cmd === "z") {
      /* close */
    }
  }
  const bw = maxX - minX;
  const bh = maxY - minY;
  if (bw <= 0 || bh <= 0) {
    throw new Error(`Degenerate wireframe bbox from path d=${d.slice(0, 80)}…`);
  }
  return [minX, minY, bw, bh];
}

function partitionIconChildren(iconRoot: SvgElement): [SvgElement[], SvgElement[]] {
  const defsNodes: SvgElement[] = [];
  const other: SvgElement[] = [];
  for (const ch of elementChildren(iconRoot)) {
    if (localTag(ch) === "defs") defsNodes.push(ch);
    else other.push(ch);
  }
  return [defsNodes, other];
}

function classHasWireframe(classAttr: string | null): boolean {
  if (!classAttr) return false;
  return classAttr.split(/\s+/).includes(WIREFRAME_CLASS);
}

function mergeIconDefsInto(
  templateDefs: SvgElement,
  defsElements: SvgElement[],
  templateDoc: SvgDocument
): void {
  const existingIds = new Set<string>();
  const collectIds = (el: SvgElement) => {
    const id = el.getAttribute("id");
    if (id) existingIds.add(id);
    for (const c of elementChildren(el)) collectIds(c);
  };
  collectIds(templateDefs);

  for (const defsEl of defsElements) {
    for (const child of [...elementChildren(defsEl)]) {
      const cid = child.getAttribute("id");
      if (cid !== null && existingIds.has(cid)) continue;
      const imported = templateDoc.importNode(child, true);
      templateDefs.appendChild(imported);
      if (cid !== null) existingIds.add(cid);
    }
  }
}

function fillSlot(
  slot: SvgElement,
  iconPath: string,
  templateDefs: SvgElement,
  templateDoc: SvgDocument
): void {
  let wire: SvgElement | null = null;
  for (const child of elementChildren(slot)) {
    if (localTag(child) !== "path") continue;
    if (classHasWireframe(child.getAttribute("class"))) {
      wire = child;
      break;
    }
  }
  if (!wire) {
    throw new Error(`Slot <g id="${slot.getAttribute("id")}"> has no ${WIREFRAME_CLASS} path.`);
  }
  const d = wire.getAttribute("d");
  if (!d) {
    throw new Error(`Wireframe path in slot ${slot.getAttribute("id")} has no d.`);
  }
  const [bx, by, bw, bh] = pathBBoxFromD(d);

  const iconDoc = parseSvgFile(iconPath);
  const iconRoot = svgDocumentElement(iconDoc);
  const [ox, oy, sw, sh] = readViewBox(iconRoot);
  if (sw <= 0 || sh <= 0) {
    throw new Error(`Invalid icon viewBox dimensions in ${iconPath}`);
  }
  const cxBox = bx + bw / 2;
  const cyBox = by + bh / 2;
  const cxIcon = ox + sw / 2;
  const cyIcon = oy + sh / 2;
  const tx = cxBox - cxIcon;
  const ty = cyBox - cyIcon;

  const [defsNodes, visual] = partitionIconChildren(iconRoot);
  if (defsNodes.length) {
    mergeIconDefsInto(templateDefs, defsNodes, templateDoc);
  }

  slot.removeChild(wire);

  const wrap = templateDoc.createElementNS(SVG_NS, "g");
  wrap.setAttribute("transform", `translate(${tx} ${ty})`);
  for (const node of visual) {
    wrap.appendChild(templateDoc.importNode(node, true));
  }
  slot.appendChild(wrap);
}

export function runPhase5(
  templatePath: PathLike,
  iconsDir: PathLike,
  outputPath: PathLike,
  options: { missing?: MissingPolicy } = {}
): { filled: number; skipped: number; missingIds: string[] } {
  const missing = options.missing ?? "skip";
  const tp = String(templatePath);
  const idir = String(iconsDir);
  const out = String(outputPath);
  if (!existsSync(tp)) throw new Error(`Template not found: ${tp}`);
  if (!existsSync(idir)) throw new Error(`Icons directory not found: ${idir}`);

  const templateDoc = parseSvgFile(tp);
  const root = svgDocumentElement(templateDoc);
  const symbols = findSymbolsGroup(root);
  const templateDefs = ensureTemplateDefs(root);
  const expected = expectedSlotIds();

  let filled = 0;
  let skipped = 0;
  const missingIds: string[] = [];

  for (const slot of [...elementChildren(symbols)]) {
    if (localTag(slot) !== "g") continue;
    const sid = slot.getAttribute("id");
    if (!sid || !expected.has(sid)) continue;
    if (!parseWeightStem(sid)) continue;

    const iconFile = join(idir, `${sid}.svg`);
    if (!existsSync(iconFile)) {
      missingIds.push(sid);
      if (missing === "fail") {
        throw new Error(`Missing icon for template slot ${sid}: ${iconFile}`);
      }
      if (missing === "warn") {
        console.warn(`Phase 5: missing icon for slot ${sid} (${iconFile}), leaving wireframe.`);
      }
      skipped++;
      continue;
    }

    fillSlot(slot, iconFile, templateDefs, templateDoc);
    filled++;
  }

  mkdirSync(dirname(out), { recursive: true });
  writeFileSync(out, elementToBytes(root));

  return { filled, skipped, missingIds };
}
